### PR TITLE
fixed adding tokenizer to hf

### PR DESCRIPTION
**Pull request by KuuCi**
# What does this PR do?

Originally the function 'write_huggingface_pretrained_from_composer_checkpoint' would not properly write the tokenizer from the composer checkpoint. This change fixes that by additionally saving the tokenizer from the composer checkpoint to the HuggingFace checkpoint. 

An additional test was implemented for this issue.

# What issue(s) does this change relate to?

- Fixes CO-1976 (https://mosaicml.atlassian.net/browse/CO-1976)

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/models/huggingface.py [hunk 1]
```diff
@@ -211,12 +307,14 @@ def hf_from_composer_checkpoint(
         Returns:
             Tuple[transformers.PreTrainedModel, Optional[transformers.PreTrainedTokenizer]]: The loaded HuggingFace model and (if present) tokenizer
         """
+        '''
```
Comment (dakinggg) [inline composer/models/huggingface.py:310]: I think these quotes should be removed?

File: composer/models/huggingface.py [hunk 2]
```diff
@@ -211,12 +307,14 @@ def hf_from_composer_checkpoint(
...
         try:
             import transformers
         except ImportError as e:
             raise MissingConditionalImportError(extra_deps_group='nlp',
                                                 conda_package='transformers',
                                                 conda_channel='conda-forge') from e
+        '''
```
Comment (dakinggg) [inline composer/models/huggingface.py:317]: same

File: tests/models/test_hf_model.py [hunk 3]
```diff
@@ -30,6 +30,26 @@
 from tests.loggers.test_remote_uploader_downloader import DummyObjectStore
 
 
+def test_hf_tokenizer_save(tmp_path: Path, tiny_bert_model, tiny_bert_tokenizer):
+    transformers = pytest.importorskip('transformers')
+
+    trainer = get_lm_trainer(tiny_bert_model, tiny_bert_tokenizer, str(tmp_path), is_conditional_generation=True)
```
Comment (dakinggg) [inline tests/models/test_hf_model.py:36]: doesn't really matter for the purposes of this test, but `is_conditional_generation` should be `False` for a bert masked language modeling model :)

File: composer/models/huggingface.py [hunk 4]
```diff
@@ -613,6 +638,12 @@ def write_huggingface_pretrained_from_composer_checkpoint(
 
     composer_state_dict = safe_torch_load(local_checkpoint_save_location)
 
+    # load tokenizer
+    hf_state = composer_state_dict['state']['integrations']['huggingface']
+    hf_tokenizer = HuggingFaceModel.write_huggingface_tokenizer_from_composer_checkpoint(hf_state)
```
Comment (dakinggg) [inline composer/models/huggingface.py:643]:
```suggestion
    hf_tokenizer = HuggingFaceModel.load_huggingface_tokenizer_from_saved_state(hf_state)
```

File: composer/models/huggingface.py [hunk 5]
```diff
@@ -139,6 +139,102 @@ def __init__(self,
 
         self.dummy_forward_called = False
 
+    @staticmethod
+    def write_huggingface_tokenizer_from_composer_checkpoint(
```
Comment (dakinggg) [inline composer/models/huggingface.py:143]: lets rename this function to `load_huggingface_tokenizer_from_saved_state` to more accurately describe what is does. Also, please add a docstring briefly describing what the function does, and the arguments (you can follow the format of other docstrings in this file)

File: composer/models/huggingface.py [hunk 6]
```diff
@@ -139,6 +139,102 @@ def __init__(self,
...
+                hf_tokenizer.name_or_path = hf_tokenizer_state['tokenizer_config']['content'].get('name_or_path', '')
+                hf_tokenizer.init_kwargs['name_or_path'] = hf_tokenizer.name_or_path
+
+                # for an unknown reason this key is missing when loading the saved tokenizer, but present with a value of None
+                # for the original tokenizer, so we default it to None
+                hf_tokenizer.init_kwargs['tokenizer_file'] = hf_tokenizer.init_kwargs.get('tokenizer_file', None)
+        return hf_tokenizer
+
+    @staticmethod
+    def write_huggingface_model_from_composer_checkpoint(
```
Comment (dakinggg) [inline composer/models/huggingface.py:188]: same, suggest renaming this to `load_huggingface_model_from_saved_state`, and please add a docstring

**Review by dakinggg — COMMENTED** [review]
Basically looks good, left a few comments!

File: composer/models/huggingface.py [hunk 7]
```diff
@@ -139,6 +139,117 @@ def __init__(self,
 
         self.dummy_forward_called = False
 
+    @staticmethod
+    def load_huggingface_tokenizer_from_saved_state(
+            hf_state: Dict[str, Any]) -> Optional[transformers.PreTrainedTokenizer]:
+        """A helper function that loads a HuggingFace tokenizer from a loaded in hf state.
+
+        Args:
+            hf_state (Dict[str, Any]): A loaded state from a checkpoint save location of a hf model.
```
Comment (dakinggg) [inline composer/models/huggingface.py:148]:
```suggestion
            hf_state (Dict[str, Any]): HF state loaded from a Composer checkpoint.
```

File: composer/models/huggingface.py [hunk 8]
```diff
@@ -139,6 +139,117 @@ def __init__(self,
...
+                # for the original tokenizer, so we default it to None
+                hf_tokenizer.init_kwargs['tokenizer_file'] = hf_tokenizer.init_kwargs.get('tokenizer_file', None)
+        return hf_tokenizer
+
+    @staticmethod
+    def load_huggingface_model_from_saved_state(
+            hf_state: Dict[str, Any], loaded_state_dict: Dict[str, Dict[str, Dict[str, Dict[str, Any]]]],
+            model_instantiation_class: type | str | None,
+            model_config_kwargs: Dict[str, Any] | None) -> transformers.PreTrainedModel:
+        """A helper function that loads a HuggingFace model from a loaded in hf state.
```
Comment (dakinggg) [inline composer/models/huggingface.py:199]:
```suggestion
        """A helper function that loads a HuggingFace model class from a loaded in hf state.
```

File: composer/models/huggingface.py [hunk 9]
```diff
@@ -139,6 +139,117 @@ def __init__(self,
...
+
+        Args:
+            hf_state (Dict[str, Any]): A loaded state from a checkpoint save location of a hf model.
```
Comment (dakinggg) [inline composer/models/huggingface.py:202]:
```suggestion
            hf_state (Dict[str, Any]): HF state loaded from a Composer checkpoint.
```

File: composer/models/huggingface.py [hunk 10]
```diff
@@ -139,6 +139,117 @@ def __init__(self,
...
+        return hf_tokenizer
+
+    @staticmethod
+    def load_huggingface_model_from_saved_state(
+            hf_state: Dict[str, Any], loaded_state_dict: Dict[str, Dict[str, Dict[str, Dict[str, Any]]]],
+            model_instantiation_class: type | str | None,
+            model_config_kwargs: Dict[str, Any] | None) -> transformers.PreTrainedModel:
+        """A helper function that loads a HuggingFace model from a loaded in hf state.
+
+        Args:
```
Comment (dakinggg) [inline composer/models/huggingface.py:201]: Can you add descriptions of the other args too? You can mostly copy the descriptions from `hf_from_composer_checkpoint`

**Review by dakinggg — APPROVED** [review]
LGTM! Once you resolve the last couple comments, feel free to merge :)
